module/zcommon/zfs_fletcher.c

```diff
@@ -245,7 +214,107 @@ fletcher_4_incremental_byteswap(const void *buf, uint64_t size,
 	ZIO_SET_CHECKSUM(zcp, a, b, c, d);
 }
 
+static const struct fletcher_4_calls fletcher_4_generic_calls = {
+	.init = fletcher_4_generic_init,
+	.compute = fletcher_4_generic,
+	.compute_byteswap = fletcher_4_generic_byteswap,
+	.name = "generic"
+};
+
+static const struct fletcher_4_calls *chosen = &fletcher_4_generic_calls;
+
+void
+fletcher_4_native(const void *buf, uint64_t size, zio_cksum_t *zcp)
+{
+	chosen->init(zcp);
+	chosen->compute(buf, size, zcp);
+	if (chosen->fini != NULL)
+		chosen->fini(zcp);
+}
+
+void
+fletcher_4_byteswap(const void *buf, uint64_t size, zio_cksum_t *zcp)
+{
+	chosen->init(zcp);
+	chosen->compute_byteswap(buf, size, zcp);
+	if (chosen->fini != NULL)
+		chosen->fini(zcp);
+}
+
+void
+fletcher_4_incremental_native(const void *buf, uint64_t size,
+    zio_cksum_t *zcp)
+{
+	fletcher_4_generic(buf, size, zcp);
+}
+
+void
+fletcher_4_incremental_byteswap(const void *buf, uint64_t size,
+    zio_cksum_t *zcp)
+{
+	fletcher_4_generic_byteswap(buf, size, zcp);
+}
+
+
 #if defined(_KERNEL) && defined(HAVE_SPL)
+
+extern struct fletcher_4_calls fletcher_4_avx2_calls;
+
+static const struct fletcher_4_calls *fletcher_4_algos[] = {
+	&fletcher_4_generic_calls,
+#if defined(HAVE_AVX) && defined(HAVE_AVX2)
+	&fletcher_4_avx2_calls,
+#endif
+};
+
+#define	BENCH_SIZE 4096
+#define	ZFS_ARRAY_SIZE(a) (sizeof (a) / sizeof (*a))
+#define	kernel_cpu_relax() do {} while (0)
+
+/* cant use allocation methods from zfs module! */
+static char databuf[BENCH_SIZE];
+
+void
+fletcher_4_init(void)
+{
+	unsigned long bestperf = 0;
+	const unsigned int bits = 4;
+	int i;
+
+	for (i = 0; i < ZFS_ARRAY_SIZE(fletcher_4_algos); i++) {
+		const struct fletcher_4_calls *algo = fletcher_4_algos[i];
+		unsigned long perf = 0;
+		clock_t j0, j1;
+		zio_cksum_t zc;
+
+		if (algo->valid != NULL && !algo->valid())
+			continue;
+
+		kpreempt_disable();
+		j0 = ddi_get_lbolt();
+		while ((j1 = ddi_get_lbolt()) == j0) {
+			kernel_cpu_relax();
+		}
+
+		algo->init(&zc);
+		while (ddi_time_before(ddi_get_lbolt(), j1 + (1 << bits))) {
+			algo->compute(databuf, PAGE_SIZE, &zc);
```

Comment:
I  introduced BENCH_SIZE, but missed to exchange PAGE_SIZE here.

Also, I opted for `gethrtime()` interface for benchmark. [see here](https://github.com/ironMann/zfs/blob/wip-simd/module/zfs/vdev_raidz_math.c#L415-L422) It's much simpler, and works from runtime also.